<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Connection-à-la-bhp" data-toc-modified-id="Connection-à-la-bhp-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Connection à la bhp</a></span></li><li><span><a href="#Exploration" data-toc-modified-id="Exploration-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Exploration</a></span><ul class="toc-item"><li><span><a href="#Types-d'objets-rattachés-au-projet" data-toc-modified-id="Types-d'objets-rattachés-au-projet-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Types d'objets rattachés au projet</a></span></li><li><span><a href="#CoAc-:-nombre-créés-par-G.-Bodé" data-toc-modified-id="CoAc-:-nombre-créés-par-G.-Bodé-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>CoAc : nombre créés par G. Bodé</a></span></li><li><span><a href="#Liste-des-CoAc-attachés-au-projet-avec-création-d'un-fichier" data-toc-modified-id="Liste-des-CoAc-attachés-au-projet-avec-création-d'un-fichier-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Liste des CoAc attachés au projet avec création d'un fichier</a></span></li><li><span><a href="#CoAc-:-distribution-des-types-de-CoAc-associés-au-projet" data-toc-modified-id="CoAc-:-distribution-des-types-de-CoAc-associés-au-projet-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>CoAc : distribution des types de CoAc associés au projet</a></span></li><li><span><a href="#CoAc-:-liste-des-informations-associées-aux-CoAc-du-projet" data-toc-modified-id="CoAc-:-liste-des-informations-associées-aux-CoAc-du-projet-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>CoAc : liste des informations associées aux CoAc du projet</a></span><ul class="toc-item"><li><span><a href="#Types-de-rôles-joués-par-les-CoAc" data-toc-modified-id="Types-de-rôles-joués-par-les-CoAc-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>Types de rôles joués par les CoAc</a></span></li><li><span><a href="#Liste-des-informations-liées-aux-CoAc-associés-au-projet." data-toc-modified-id="Liste-des-informations-liées-aux-CoAc-associés-au-projet.-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>Liste des informations liées aux CoAc associés au projet.</a></span></li><li><span><a href="#Liste-des-créateurs-des-informations" data-toc-modified-id="Liste-des-créateurs-des-informations-2.5.3"><span class="toc-item-num">2.5.3&nbsp;&nbsp;</span>Liste des créateurs des informations</a></span></li></ul></li><li><span><a href="#Informations-création" data-toc-modified-id="Informations-création-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Informations création</a></span><ul class="toc-item"><li><span><a href="#Liste-des-TyIn30,-création-d'un-acteur-collectif." data-toc-modified-id="Liste-des-TyIn30,-création-d'un-acteur-collectif.-2.6.1"><span class="toc-item-num">2.6.1&nbsp;&nbsp;</span>Liste des TyIn30, création d'un acteur collectif.</a></span></li><li><span><a href="#Distribution-des-dates" data-toc-modified-id="Distribution-des-dates-2.6.2"><span class="toc-item-num">2.6.2&nbsp;&nbsp;</span>Distribution des dates</a></span></li></ul></li></ul></li></ul></div>

# Premières explorations du corpus du projet Etablissements d'enseignement technique

Importation des bibliothèques et modules nécessaires


In [1]:
import os
import pathlib
from pathlib import Path
import matplotlib
from matplotlib import pyplot as plt
import psycopg2
import getpass
import pandas as pd
import seaborn as sns

## Connection à la bhp

Saisie du mot de passe, le même que celui utilisé pour son compte utilisateur de la BHP.

In [2]:
password = getpass.getpass()

········


Après enregistrement du notebook sur sa machine compléter sur la version récupérée sur sa machine l'hôte, le port, votre login tels qu'ils vous ont été fournis.

In [3]:
conn = psycopg2.connect(host = "hôte",
port = "port",
user = "login",
password = password,
database = "symogih") 

In [4]:
cursor = conn.cursor()
print("PostgreSQL server information")
print(conn.get_dsn_parameters(), "\n")

PostgreSQL server information
{'user': 'pvernus', 'channel_binding': 'prefer', 'dbname': 'symogih', 'host': 'bhp-publi.ish-lyon.cnrs.fr', 'port': '9375', 'tty': '', 'options': '', 'sslmode': 'prefer', 'sslcompression': '0', 'ssl_min_protocol_version': 'TLSv1.2', 'gssencmode': 'prefer', 'krbsrvname': 'postgres', 'target_session_attrs': 'any'} 



In [5]:
cursor.execute("SELECT version();")
record = cursor.fetchone()
print("you are connected to - ", record, "\n")

you are connected to -  ('PostgreSQL 9.4.18 on x86_64-unknown-linux-gnu (Debian 9.4.18-2.pgdg70+1), compiled by gcc (Debian 4.7.2-5) 4.7.2, 64-bit',) 



## Exploration

### Types d'objets rattachés au projet

In [6]:
sql = """
SELECT substring(fk_object from 1 for 4), count(*) nombre
  FROM bhp.v_associate_project
  WHERE fk_collective_actor = 15789
  GROUP BY substring(fk_object from 1 for 4)
  ORDER BY nombre desc
"""
cursor.execute(sql)
query = cursor.fetchall()
df = pd.DataFrame(query, columns=["type_objet", "freq"])
print(df)

  type_objet  freq
0       CoAc  5388
1       MaOb  1358
2       AbOb   632
3       Actr   141
4       SoCh    35
5       Info    12
6       Bibl     9
7       TyIn     6
8       DiOb     3
9       NaPl     1


### CoAc : nombre créés par G. Bodé

In [7]:
sql = """
SELECT count(*)
  FROM bhp.v_collective_actor
  WHERE creator = 74
"""
cursor.execute(sql)
query = cursor.fetchall()
print(query)


[(5436,)]


### Liste des CoAc attachés au projet avec création d'un fichier

In [8]:
sql = """
SELECT concat_coac, concat_standard_name, notes_begin, certainty_begin, begin_year
FROM bhp.v_collective_actor
WHERE trim(concat_coac)
IN (
SELECT trim(fk_object)
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%'
)
ORDER BY begin_year
"""
cursor.execute(sql)
query = cursor.fetchall()
df = pd.DataFrame(query, columns=["pk", "standard_name", "annee_deb_qualif", "annee_deb_certitude", "annee_deb"])
df.to_excel("liste_coac.xlsx")
df.head()

pk                      standard_name annee_deb_qualif  \
0  CoAc22392                  Diocèse de Tarbes                2   
1  CoAc16176       École d’art du pont (Toulon)                2   
2  CoAc18859      Ecole d'hydrographie (Dieppe)                2   
3  CoAc18858    Ecole d'hydrographie (Le Havre)                2   
4  CoAc18863  Ecole d'hydrographie (Saint-Malo)                2   

  annee_deb_certitude  annee_deb  
0                   1        506  
1                   1       1640  
2                   1       1665  
3                   1       1665  
4                   1       1669

Problème il n'y a que 5387 résultats au lieu de 5388 comme dans le résultat de la requête précédente. Pourquoi ?
En fait le CoAc18825 a été associé deux fois au projet comme le montre la requête suivante.

**Il n'y a donc que 5387 CoAc associés**.

In [9]:
sql = """
SELECT fk_object, count(*) freq
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%'
GROUP BY fk_object
ORDER BY freq desc
"""
cursor.execute(sql)
query = cursor.fetchall()
df = pd.DataFrame(query, columns=["coac", "freq"])
df.head()

coac  freq
0  CoAc18825     2
1  CoAc17775     1
2  CoAc16810     1
3  CoAc17802     1
4  CoAc19067     1

On aura donc la liste des CoAc associés au projet avec la requête suivante : 

`SELECT DISTINCT fk_object
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%'`

### CoAc : distribution des types de CoAc associés au projet

In [10]:
sql = """
SELECT fk_abob_type_collective_actor, count(*) freq
  FROM bhp.v_collective_actor
  WHERE concat_coac IN 
  (SELECT DISTINCT fk_object
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%'
)
GROUP BY fk_abob_type_collective_actor
ORDER BY freq desc
"""
cursor.execute(sql)
query = cursor.fetchall()
print(query)

[(1051, 5387)]


Tous les acteurs collectis sont du même type : 1051, collectivité structurée

### CoAc : liste des informations associées aux CoAc du projet

D'abord, quelles sont les informations qui sont associées aux CoAc du projet ?

In [11]:
sql = """
SELECT fk_information, count(*) freq
  FROM bhp.v_information_role
  WHERE fk_associated_object IN (
  SELECT DISTINCT fk_object
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%'
)
GROUP BY fk_information
ORDER BY freq desc;
"""
cursor.execute(sql)
query = cursor.fetchall()
df = pd.DataFrame(query, columns=["fk_info", "freq"])
df[df["freq"] > 1]

fk_info  freq
0   158501     2
1   151977     2
2   134127     2
3   147061     2
4   157189     2

Il y a donc 5 informations auxquels le CoAc a été associé deux fois => à corriger.
Combien y a-t-il d'informations ?

In [12]:
len(df)

20985

#### Types de rôles joués par les CoAc

In [13]:
sql = """
SELECT fk_type_role, count(*) freq
  FROM bhp.v_information_role
  WHERE fk_associated_object IN (
  SELECT DISTINCT fk_object
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%'
)
GROUP BY fk_type_role
ORDER BY freq desc;
"""
cursor.execute(sql)
query = cursor.fetchall()
df = pd.DataFrame(query, columns=["type_role", "freq"])
df

type_role  freq
0        21.0  9126
1        49.0  5368
2        12.0  4579
3        41.0  1822
4       133.0    36
5        98.0    18
6        54.0    16
7         NaN    14
8       174.0     3
9       131.0     2
10        8.0     2
11       58.0     1
12       50.0     1
13       37.0     1
14      195.0     1

14 roles ne sont pas typés. Voici la liste des informations concernées 

In [14]:
sql = """
SELECT fk_information
  FROM bhp.v_information_role
  WHERE fk_associated_object IN (
  SELECT DISTINCT fk_object
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%'
)
AND fk_type_role is null
"""
cursor.execute(sql)
query = cursor.fetchall()
print(query)

[(88702,), (88749,), (73640,), (78636,), (87932,), (88344,), (78362,), (75023,), (49841,), (73655,), (78193,), (52165,), (52238,), (53896,)]


#### Liste des informations liées aux CoAc associés au projet.

In [18]:
sql = """
SELECT info.pk_information, info.concat_info, info.notes, type_info.concat_standard_name, 
      info.fk_inherit_info_sourcing, info.creator, info.info_label, info.standard_date, info.standard_date_type, 
       info.info_subject
FROM bhp.v_information info
JOIN bhp.v_type_information type_info
ON info.fk_type_information = type_info.pk_type_information
WHERE pk_information IN
(
SELECT DISTINCT fk_information
  FROM bhp.v_information_role
  WHERE fk_associated_object IN (
  SELECT DISTINCT fk_object
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%')
)
"""
cursor.execute(sql)
query = cursor.fetchall()
df = pd.DataFrame(query, columns=["pk_information", "concat_info", "notes", "fk_type_information", 
      "fk_inherit_info_sourcing", "creator", "info_label", "standard_date", "standard_date_type", 
       "info_subject"])
p = Path("../tableaux/listes_informations_liees_aux_coac.xlsx")
df.to_csv(p)
df.head()

pk_information concat_info notes  \
0          153566  Info153566  None   
1          155132  Info155132  None   
2          155028  Info155028  None   
3          141498  Info141498  None   
4           87832   Info87832         

                             fk_type_information fk_inherit_info_sourcing  \
0  Activité d'enseignement d'un acteur collectif                     None   
1                    Caractéristiques d'un objet                     None   
2                    Caractéristiques d'un objet                     None   
3                 Création d'un acteur collectif                     None   
4                          Qualité (obtenir une)                     None   

   creator                                         info_label standard_date  \
0     74.0  Ecole pratique d'agriculture (Plouguernével); ...    1909-04-16   
1     74.0  Ferme-école de Varincourt (Laneuvelotte); Inte...    1849-04-03   
2     74.0  Ferme-école d'Ollwiller (Wuenheim); Internat -...    1849-05-14   
3     74.0  Section commerciale de l'école primaire supéri...          1931   
4     11.0  Marrel, Jean - Diplômé des Arts et Métiers (Cl...          1906   

  standard_date_type info_subject  
0      Date de début         None  
1      Date de début         None  
2      Date de début         None  
3        Date unique         None  
4         Date début         None

Distribution des types d'information au sein de cette liste :
30 : création
41 : statut
166 : activité d'enseignement d'un CoAc
148 : caractéristique d'un objet
11 : qualité (obtenir)
7 : fonction (exercice)

In [6]:
df["fk_type_information"].value_counts()

Création d'un acteur collectif                   5372
Statut                                           5369
Activité d'enseignement d'un acteur collectif    4598
Caractéristiques d'un objet                      3243
Qualité (obtenir une)                            2093
Fonction (exercice de la )                        226
Formation (suivre une)                             36
Fin de l'existence d'un acteur collectif           15
Enseignement                                       15
Évènement                                          11
Nombre                                              6
Rattachement à un acteur collectif                  1
Name: fk_type_information, dtype: int64

#### Liste des créateurs des informations

In [29]:
df["creator"].value_counts()

74.0     18740
12.0      1240
11.0       881
120.0       47
86.0        37
73.0        15
50.0         6
20.0         4
55.0         3
18.0         3
26.0         3
2.0          2
51.0         1
1.0          1
Name: creator, dtype: int64

### Informations création

#### Liste des TyIn30, création d'un acteur collectif.

In [20]:
sql = """
SELECT info.pk_information, info.concat_info, info.notes, type_info.concat_standard_name, 
      info.fk_inherit_info_sourcing, info.creator, info.info_label, info.standard_date, info.standard_date_type, 
       info.info_subject
FROM bhp.v_information info
JOIN bhp.v_type_information type_info
ON info.fk_type_information = type_info.pk_type_information
WHERE pk_information IN
(
SELECT DISTINCT fk_information
  FROM bhp.v_information_role
  WHERE fk_associated_object IN (
  SELECT DISTINCT fk_object
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%')
)
AND info.fk_type_information = 30
"""
cursor.execute(sql)
query = cursor.fetchall()
df = pd.DataFrame(query, columns=["pk_information", "concat_info", "notes", "fk_type_information", 
      "fk_inherit_info_sourcing", "creator", "info_label", "standard_date", "standard_date_type", 
       "info_subject"])
p = Path("../tableaux/type_info_30.xlsx")
df.to_csv(p)
len(df)

5372

#### Distribution des dates

In [8]:
sql = """
SELECT info.standard_date, count(*) freq
FROM bhp.v_information info
JOIN bhp.v_type_information type_info
ON info.fk_type_information = type_info.pk_type_information
WHERE pk_information IN
(
SELECT DISTINCT fk_information
  FROM bhp.v_information_role
  WHERE fk_associated_object IN (
  SELECT DISTINCT fk_object
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%')
)
AND info.fk_type_information = 30
GROUP BY info.standard_date
ORDER BY freq desc
"""
cursor.execute(sql)
query = cursor.fetchall()
df = pd.DataFrame(query, columns=["standard_date", "freq"])
p = Path("../tableaux/distribution_type_dates.xlsx")
df.to_csv(p)
df.head(25)

standard_date  freq
0         1931 p   427
1         1930 p   336
2         1928 p   331
3         1933 p   323
4         1920 p   173
5         1895 p   125
6         1921 p    92
7     1935-09-17    86
8         1922 p    68
9         1938 p    61
10          1920    48
11        1910 p    43
12        1936 p    36
13        1900 p    36
14        1929 p    35
15        1911 p    34
16        1939 p    33
17    1937-07-26    32
18        1934 p    32
19    1920-05-21    31
20        1912 p    30
21        1937 p    29
22        1799 p    28
23        1925 p    28
24        1899 p    27

Beaucoup de dates sont postulées, combien y en a-t-il ?

In [22]:
sql = """
SELECT count(*)
FROM bhp.v_information info
JOIN bhp.v_type_information type_info
ON info.fk_type_information = type_info.pk_type_information
WHERE pk_information IN
(
SELECT DISTINCT fk_information
  FROM bhp.v_information_role
  WHERE fk_associated_object IN (
  SELECT DISTINCT fk_object
FROM bhp.v_associate_project
WHERE fk_collective_actor = 15789
AND fk_object LIKE 'CoAc%')
)
AND info.fk_type_information = 30
AND info.standard_date LIKE '%p%'
"""
cursor.execute(sql)
query = cursor.fetchall()
print(query[0][0])

2866


Donc 2866 dates de création sont postulées sur 5372 soit plus de la moitié.